### ISSUE TITLE
Fish Shell adds 4m characters to completion in Ansi-term Emacs

### ISSUE THREAD

**Issue by jaxspades** [fish-shell/fish-shell#2441]
I just recently installed the fish shell, and so far, I love the completion. Pretty much exactly what the shell has needed since forever! It works amazingly in iTerm on the Mac, but when I use it on ansi-term for my emacs client, it adds the '4m' characters between each character I type. So, `git status` becomes `git s4mt4ma4mt4mu4ms`. I think that's also why I don't get completion examples in ansi-term, like I would in iTerm.

All of my other shells work, but then again, they don't use completion, so it might be that ansi-term doesn't support the terminal based completion. Anyways, I'm adding the issue here to see if anyone else who uses Fish Shell, also uses it from ansi-term and whether or not this is a known issue with a resolution or if I potentially have an oddly broken setup.

**Comment by jaxspades** [comment]
A bit more information, this only happens on commands that auto-complete, like linux commands and git. I noticed that grunt, npm, svn and some other commands that do not auto-complete do not have this issue.

**Comment by faho** [comment]
Which fish version is this? 

>this only happens on commands that auto-complete

What do you mean by auto-complete? The suggestions or what you get after pressing TAB?

What do `echo $TERM` and `tput colors` say?

**Comment by jaxspades** [comment]
The suggestions get the weird characters.

echo $TERM - xterm-256color
tput colors - 256

emacs version - 24.5.1
fish version - 2.2.0

**Comment by faho** [comment]
In that case, I don't think ansi-term supports 256 colors - it certainly doesn't handle them like xterm, so setting TERM to "xterm-256color" is _wrong_.

I haven't been able to find a magic invocation to make it do 256 colors, so: Sorry. (Well, [xterm-color.el](https://github.com/atomontage/xterm-color) is supposed to do this, but I haven't found a way to make it work).

**Comment by jaxspades** [comment]
How does that cause my suggestions to have extra characters?

**Comment by faho** [comment]
The way all this "shell in a terminal" stuff works is that they communicate via characters. The shell sends characters, like text, but also specific sequences to tell the terminal "this is supposed to be green".

When the terminal doesn't recognize the code, it will probably do a wrong thing, which may be that it prints the characters instead of doing the thing the characters are supposed to signify.

When you check `infocmp xterm-256color`, it prints a list of "things" with the escape sequences that xterm-256color thinks mean those things. `infocmp xterm-256color | grep 4m` tells me it's either "smul" or "rmul", which `man terminfo` tells me is "start underline mode" or "exit underline mode".

In short: Fish tells the terminal to start or stop underlining in the particular dialect that xterm understands because the terminal identifies as xterm (that's what the TERM variable is for). If that terminal doesn't behave like xterm, then there's nothing we can do - we can't just start guessing.

**Comment by zanchey** [comment]
So many programs lie about being Xterm that wherever possible we should add workarounds, but there's not much fish can do here - there's no way of telling how many colours are supported except through terminfo.

**Comment by faho** [comment]
Exactly - for e.g. keycodes, we can accept all of them unless they contradict each other. But for this - we can't send all sequences because that would break _all_ terminals. They always contradict each other.

I think I've already said that I rather dislike this entire situation - I still don't get why you'd write your own set of sequences for existing things (i.e. why not just behave like xterm _unless you're extending it_). It almost seems like incompatibility for the sake of it.

**Comment by faho** [comment]
@AriaFallah: I don't get that in my emacs 24.5.1 - do you have any customization for term or ansi-term? What's your config.fish and universal variables (`set -U` or the contents of ~/.config/fish/fishd.*)?

(I haven't run it on a pristine fish 2.2.0 - I'll try that now)

**Comment by AriaFallah** [comment]
I don't have any term customization and I don't have much in my fish config:

config.fish
```
# Set the greeting for when opening a new terminal.
set -gx fish_greeting 'The world is at the tips of your fingers :D'

# Set variables necessary for PATH
set -gx NVM_DIR '/Users/Developer/.nvm'
set -gx GOPATH '/Users/Developer/workspace/go'
set -gx PATH $PATH /usr/local/sbin $NVM_DIR $GOPATH /Users/Developer/anaconda2/bin

# activate nvm
bass source ~/.nvm/nvm.sh ';' nvm use stable
```

My prompt also just in case that matters:

```
function fish_prompt --description 'Write out the prompt'
	set -l last_status $status

	if not set -q __fish_prompt_normal
		set -g __fish_prompt_normal (set_color normal)
	end

	# PWD
	set_color $fish_color_cwd
	echo -n (prompt_pwd)
	set_color normal

	printf '%s ' (__fish_git_prompt)

	if not test $last_status -eq 0
	set_color $fish_color_error
	end

	echo -n '~> '

	set_color normal
end
```

Hopefully that'll give you some leads.

**Comment by pickfire** [comment]
@AriaFallah Hi, you can actually use syntax highlighting by adding the following at the beginning of code:

    ```fish

**Comment by AriaFallah** [comment]
Just want to say this is solved by following the instructions you can find [here](http://emacs.stackexchange.com/questions/7372/stray-trailing-4m-before-prompt-with-zsh-in-m-x-ansi-term)

The problem is caused by having some problem in `eterm-colors.ti` so you need to recompile it.

**Comment by krader1961** [comment]
This is also documented at http://www.emacswiki.org/emacs/AnsiTermHints. Short answer: ansi-term implements a subset of the capabilities supported by xterm256 so you can't tell fish that you're using a xterm256; you have to set TERM to eterm-color and install the corresponding terminfo file. As @zanchey said it would be nice if fish could auto-detect the capabilities and do the right thing but that isn't going to be possible any time soon (or ever in all likelihood).

**Comment by edrex** [comment]
It looks like fish is ignoring `$TERMINFO`, as suggested [here](http://emacs.stackexchange.com/a/7402/10277)

From within ansi-term, I ran:

```
tic -o ~/.terminfo $TERMINFO/e/eterm-color.ti
```

and now fish inside ansi-term behaves acceptable. Why is this needed?

**Comment by faho** [comment]
>and now fish inside ansi-term behaves acceptably. Why is this needed?

Fish _does_ respect that variable - you can confirm with e.g. `strace`.

It's possible the curses your fish was compiled with can't read it, which recompilation fixes. Or something.

It's telling that all reports for this have been on OSX as far as I can tell.
